From 21a43523914d15e8cac3b8b6458c0a3ad5f9a192 Mon Sep 17 00:00:00 2001
From: Ricky Stewart <rstewart@mozilla.com>
Date: Thu, 5 Dec 2019 18:56:39 +0000
Subject: [PATCH] Bug 1601407 - Update gfx/graphite2/src/moz.build to
 additionally build a wasm sandboxed library
 r=firefox-build-system-reviewers,mshal

Differential Revision: https://phabricator.services.mozilla.com/D56024

--HG--
extra : moz-landing-system : lando

diff --git a/build/templates.mozbuild b/build/templates.mozbuild
index f130e2b761fd..4aa1825560f4 100644
--- a/build/templates.mozbuild
+++ b/build/templates.mozbuild
@@ -168,3 +168,9 @@ def ForceInclude(*headers):
 
 include('gecko_templates.mozbuild')
 include('test_templates.mozbuild')
+
+
+@template
+def DefineAndWasmDefine(k, v):
+    DEFINES[k] = v
+    WASM_DEFINES[k] = v
diff --git a/gfx/graphite2/src/moz.build b/gfx/graphite2/src/moz.build
index 86791b641053..a6167943e43c 100755
--- a/gfx/graphite2/src/moz.build
+++ b/gfx/graphite2/src/moz.build
@@ -13,17 +13,8 @@ EXPORTS.graphite2 += [
     '../include/graphite2/Types.h',
 ]
 
-if CONFIG['CC_TYPE'] in ('clang', 'gcc'):
-    UNIFIED_SOURCES += [
-        'direct_machine.cpp'
-    ]
-else:
-    UNIFIED_SOURCES += [
-        'call_machine.cpp'
-    ]
-
 # This should contain all of the _SOURCES from files.mk, except *_machine.cpp
-UNIFIED_SOURCES += [
+all_sources = [
     '../geckoextra/src/GraphiteExtra.cpp',
     'CmapCache.cpp',
     'Code.cpp',
@@ -56,22 +47,39 @@ UNIFIED_SOURCES += [
     'UtfCodec.cpp',
 ]
 
+if CONFIG['CC_TYPE'] in ('clang', 'gcc'):
+    UNIFIED_SOURCES += [
+        'direct_machine.cpp'
+    ]
+else:
+    UNIFIED_SOURCES += [
+        'call_machine.cpp'
+    ]
+UNIFIED_SOURCES += all_sources
+
+if CONFIG['MOZ_WASM_SANDBOXING_GRAPHITE']:
+    SANDBOXED_WASM_LIBRARY_NAME = 'graphitewasm'
+    WASM_SOURCES += [
+        'call_machine.cpp',
+    ]
+    WASM_SOURCES += all_sources
+
 # tell graphite2 not to export symbols, we'll be linking it directly with
 # thebes
-DEFINES['GRAPHITE2_STATIC'] = True
+DefineAndWasmDefine('GRAPHITE2_STATIC', True)
 
 # We allow warnings for third-party code that can be updated from upstream.
 AllowCompilerWarnings()
 
 FINAL_LIBRARY = 'gkmedias'
 
-DEFINES['PACKAGE_VERSION'] = '"moz"'
-DEFINES['PACKAGE_BUGREPORT'] = '"http://bugzilla.mozilla.org/"'
+DefineAndWasmDefine('PACKAGE_VERSION', '"moz"')
+DefineAndWasmDefine('PACKAGE_BUGREPORT', '"http://bugzilla.mozilla.org/"')
 
 # disable features we don't need in the graphite2 code, to reduce code size
 for var in ('GRAPHITE2_NFILEFACE', 'GRAPHITE2_NTRACING', 'GRAPHITE2_NSEGCACHE'):
-    DEFINES[var] = True
+    DefineAndWasmDefine(var, True)
 
 # provide a custom header that overrides malloc() and friends,
-# to ensure safe OOM handling
+# to ensure safe OOM handling. Don't do this in the wasm sandbox.
 DEFINES['GRAPHITE2_CUSTOM_HEADER'] = '"MozGrMalloc.h"'
-- 
2.25.1

